src/geom/point.jl

```diff
@@ -47,34 +47,45 @@ function render(geom::PointGeometry, theme::Gadfly.Theme, aes::Gadfly.Aesthetics
     Gadfly.assert_aesthetics_equal_length("Geom.point", aes, :x, :y)
 
     default_aes = Gadfly.Aesthetics()
-    default_aes.shape = Function[Shape.circle]
-    default_aes.color = discretize_make_ia(RGBA{Float32}[theme.default_color])
+    default_aes.shape = Function[theme.point_shapes[1]]
+    default_aes.color = RGBA{Float32}[theme.default_color]
     default_aes.size = Measure[theme.point_size]
-    default_aes.alpha = [theme.alphas[1]]
+    default_aes.alpha = Float64[theme.alphas[1]]
     aes = inherit(aes, default_aes)
 
-    if eltype(aes.size) <: Int
+    aes_size = if eltype(aes.size) <: Int
       size_min, size_max = extrema(aes.size)
       size_range = size_max - size_min
       point_size_range = theme.point_size_max - theme.point_size_min
-      interpolate_size(x) = theme.point_size_min + (x-size_min) / size_range * point_size_range
+      theme.point_size_min .+ ((aes.size .- size_min) ./ size_range .* point_size_range)
+    else
+        aes.size
     end
 
     aes_alpha = eltype(aes.alpha) <: Int ? theme.alphas[aes.alpha] : aes.alpha
-
+    aes_shape = eltype(aes.shape) <: Function ? aes.shape : theme.point_shapes[aes.shape]
+    strokef = aes.color_key_continuous != nothing && aes.color_key_continuous ?
+                    theme.continuous_highlight_color : theme.discrete_highlight_color
+    
+    CT, ST, SZT, AT =  eltype(aes.color), eltype(aes_shape), eltype(aes_size), eltype(aes_alpha)
+    groups =   collect(Tuple{CT, SZT, ST, AT}, Compose.cyclezip(aes.color, aes_size, aes_shape, aes_alpha))
+    ug = unique(groups)
     ctx = context()
 
-    for (x, y, color, size, shape, alpha) in Compose.cyclezip(aes.x, aes.y, aes.color, aes.size, aes.shape, aes_alpha)
-        shapefun = typeof(shape) <: Function ? shape : theme.point_shapes[shape]
-        sizeval = typeof(size) <: Int ? interpolate_size(size) : size
-        strokecolor = aes.color_key_continuous != nothing && aes.color_key_continuous ?
-                    theme.continuous_highlight_color(color) :
-                    theme.discrete_highlight_color(color)
-        class = svg_color_class_from_label(aes.color_label([color])[1])
+    if length(groups)==1
```

Comment:
shouldn't this be `length(ug)` ?

Reply:
I remember thinking about this (and perhaps even testing it) but there was a reason I didn't use `length(ug)` here, and maybe it was because of my point below (`length(groups)==0)`, so using `groups` leads to simpler logic.

Reply:
But `unique` should still work on zero-length arrays? 

```julia
julia> unique([])
0-element Array{Any,1}
```

Reply:
An example:
```
p1 = plot(x=rand(1000), y=rand(1000), color=[RGB(1,0,0)], Geom.point)
p2 = plot(x=rand(1000), y=rand(1000), color=fill("a",1000), Geom.point, 
    Scale.color_discrete_manual("red"))
```
For `length(groups)`, these 2 plots are different: `p1` optimises speed (`length(groups)==1`), `p2` is slower but has full SVGJS functionality (`length(groups)>1`). Using `length(ug)` here would mean `length(ug)==1` for both `p1` and `p2`, thus not giving users the above choice.

Reply:
am i correct in thinking `p2` has a single extra call to `context`?  if so, is that really noticeably slower?

Reply:
Here are some stats:
```julia
gp, theme = Geom.point(), Theme()

aes = Gadfly.Aesthetics()
aes.x, aes.y, aes.color = rand(1000), rand(1000), [RGB(1,0,0)]
@benchmark render($gp, $theme, $aes ) samples=10

BenchmarkTools.Trial: 
  memory estimate:  95.69 KiB
  allocs estimate:  3129
  --------------
  minimum time:     128.000 μs (0.00% GC)
  median time:      144.000 μs (0.00% GC)
  mean time:        153.692 μs (0.00% GC)
  maximum time:     246.724 μs (0.00% GC)
  --------------

aes.color = fill(RGB(1,0,0), 1000)
@benchmark render($gp, $theme, $aes ) samples=10

BenchmarkTools.Trial: 
  memory estimate:  466.22 KiB
  allocs estimate:  13195
  --------------
  minimum time:     2.078 ms (0.00% GC)
  median time:      2.094 ms (0.00% GC)
  mean time:        2.122 ms (0.00% GC)
  maximum time:     2.245 ms (0.00% GC)
  --------------
```

Reply:
And compared to master. The timing is similar for both cases, so I only include the first case:
```julia
aes.x, aes.y, aes.color = rand(1000), rand(1000), [RGB(1,0,0)]
@benchmark render($gp, $theme, $aes ) samples=10

BenchmarkTools.Trial: 
  memory estimate:  4.69 MiB
  allocs estimate:  107096
  --------------
  minimum time:     6.578 ms (0.00% GC)
  median time:      6.917 ms (0.00% GC)
  mean time:        13.487 ms (49.08% GC)
  maximum time:     60.599 ms (88.65% GC)
```

Reply:
interesting.  must be the `view`s then that slow `p2` down.

my question was motivated by a desire for full SVGJS support in all cases.  the following change gets that at a cost of only 15% in performance according to my benchmarking just now, a small price in my mind:

```
diff --git a/src/geom/point.jl b/src/geom/point.jl
index 283edc32..51328d78 100644
--- a/src/geom/point.jl
+++ b/src/geom/point.jl
@@ -74,9 +74,10 @@ function render(geom::PointGeometry, theme::Gadfly.Theme, aes::Gadfly.Aesthetics
 
     if length(groups)==1
         color, size, shape, alpha = groups[1]
-        compose!(ctx, (context(),
+        class = svg_color_class_from_label(aes.color_label([color])[1])
+        compose!(ctx, (context(), (context(),
             shape(aes.x, aes.y, [size]), fill(color), stroke(strokef(color)), fillopacity(alpha),
-            svgclass("marker")))
+            svgclass("marker")), svgclass(class)))
     elseif length(groups)>1
         for g in ug
             i = findall(x->isequal(x, g), groups)
```

you might also consider putting `ug = unique(groups)` in the `elseif` clause as that's the only place it is used.

Reply:
Can you give an example of where you would want to turn off the points in a "single-color" plot? And if you wanted to do that, why not just use case 2 above?

Reply:
Geom.point combined with any other Geom of a different color (density2d, ellipse, abline you name it).  i don't think a small decrement in performance outweighs forcing the user to switch to a different grammer.